File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/menu/BottomSheetMenu.java
package com.habitrpg.android.habitica.ui.menu;

import android.content.Context;
import android.support.design.widget.BottomSheetDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.habitrpg.android.habitica.R;

import org.w3c.dom.Text;

import java.util.ArrayList;

public class BottomSheetMenu extends BottomSheetDialog implements View.OnClickListener {

    private LayoutInflater inflater;
    private Context context;
    private LinearLayout contentView;
    private BottomSheetMenuSelectionRunnable runnable;

    public BottomSheetMenu(Context context) {
        super(context);
        this.context = context;
        this.inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        this.contentView = (LinearLayout)inflater.inflate(R.layout.menu_bottom_sheet, null);
        this.setContentView(this.contentView);
    }

    public void setSelectionRunnable(BottomSheetMenuSelectionRunnable runnable) {
        this.runnable = runnable;
    }

    public void addMenuItems(BottomSheetMenuItem... menuItems) {
        for (BottomSheetMenuItem menuItem : menuItems) {
            this.addMenuItem(menuItem);
        }
    }

    public void addMenuItem(BottomSheetMenuItem menuItem) {
        View item = menuItem.inflate(this.context, this.inflater, this.contentView);
        item.setOnClickListener(this);
        this.contentView.addView(item);
    }

    public void removeMenuItem(Integer index) {
        this.contentView.removeViewAt(index);
    }

    @Override
    public void onClick(View v) {
        if (this.runnable != null) {
            Integer index = this.contentView.indexOfChild(v);
            if (index != -1) {
                this.runnable.selectedItemAt(index);
                this.dismiss();
            }
        }
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/menu/BottomSheetMenu.kt
package com.habitrpg.android.habitica.ui.menu

import android.content.Context
import android.view.View
import android.view.animation.LinearInterpolator
import android.view.animation.TranslateAnimation
import androidx.core.view.children
import com.habitrpg.android.habitica.databinding.MenuBottomSheetBinding
import com.habitrpg.android.habitica.ui.views.dialogs.HabiticaBottomSheetDialog
import com.habitrpg.common.habitica.extensions.dpToPx
import com.habitrpg.common.habitica.extensions.loadImage

class BottomSheetMenu(context: Context) : HabiticaBottomSheetDialog(context), View.OnClickListener {
    private var binding = MenuBottomSheetBinding.inflate(layoutInflater)
    private var runnable: ((Int) -> Unit)? = null

    init {
        setContentView(binding.root)
        binding.titleView.visibility = View.GONE
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        binding.menuItems.children.forEachIndexed { index, view ->
            val anim = TranslateAnimation(0f, 0f, 10f.dpToPx(context) + (5f.dpToPx(context)*index), 0f)
            anim.startOffset = 300 + (20 * index).toLong()
            anim.fillBefore = true
            anim.fillAfter = true
            anim.interpolator = LinearInterpolator()
            anim.duration = 200 + (20 * index).toLong()
            view.startAnimation(anim)
        }
    }

    fun setSelectionRunnable(runnable: (Int) -> Unit) {
        this.runnable = runnable
    }

    override fun setTitle(title: CharSequence?) {
        binding.titleView.text = title
        binding.titleView.visibility = View.VISIBLE
        binding.titleView.background = null
    }

    fun setImage(url: String) {
        binding.iconView.loadImage(url)
        binding.iconView.visibility = View.VISIBLE
    }

    fun addMenuItem(menuItem: BottomSheetMenuItem) {
        val item = menuItem.inflate(this.context, layoutInflater, this.binding.menuItems)
        item.setOnClickListener(this)
        this.binding.menuItems.addView(item)
        binding.root.requestLayout()
    }

    override fun onClick(v: View) {
        if (this.runnable != null) {
            val index = this.binding.menuItems.indexOfChild(v)
            if (index != -1) {
                runnable?.let { it(index) }
                this.dismiss()
            }
        }
    }
}
